### PR TITLE
1050: Add Redfish total_power object

### PR DESCRIPTION
**Pull request by lxwinspur**
Fixes https://github.com/ibm-openbmc/dev/issues/3620

### CODE REVIEW

File: redfish-core/lib/environment_metrics.hpp [hunk 1]
```diff
@@ -152,6 +152,83 @@ inline void handleEnvironmentMetricsHead(
                                                 std::move(respHandler));
 }
 
+inline void getPowerWatts(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                          const std::string& chassisId)
+{
+    const std::string sensorPath = "/xyz/openbmc_project/sensors";
+    constexpr std::array<std::string_view, 1> sensorInterfaces = {
+        "xyz.openbmc_project.Sensor.Value"};
+    constexpr std::array<std::string_view, 1> interfaces = {
```
Comment (cjcain) [inline redfish-core/lib/environment_metrics.hpp:161]: This line is not needed

File: redfish-core/lib/environment_metrics.hpp [hunk 2]
```diff
@@ -152,6 +152,83 @@ inline void handleEnvironmentMetricsHead(
...
+                break;
+            }
+        }
+        if (!judgment)
+        {
+            BMCWEB_LOG_DEBUG << "There is not total_power";
+            return;
+        }
+
+        constexpr std::array<std::string_view, 1> totalPowerInterfaces = {
```
Comment (cjcain) [inline redfish-core/lib/environment_metrics.hpp:191]: Could use sensorInterfaces from above.

**Review by cjcain — CHANGES_REQUESTED** [review]

File: redfish-core/lib/environment_metrics.hpp (hunk 1 above)
Comment (lxwinspur) [reply redfish-core/lib/environment_metrics.hpp]: Done

File: redfish-core/lib/environment_metrics.hpp [hunk 3]
```diff
@@ -152,6 +152,80 @@ inline void handleEnvironmentMetricsHead(
...
+    const std::string sensorPath = "/xyz/openbmc_project/sensors";
+    constexpr std::array<std::string_view, 1> sensorInterfaces = {interface};
+    dbus::utility::getSubTreePaths(
+        sensorPath, 0, sensorInterfaces,
+        [asyncResp,
+         chassisId](const boost::system::error_code& ec,
+                    const dbus::utility::MapperGetSubTreePathsResponse& paths) {
+        if (ec)
+        {
+            BMCWEB_LOG_DEBUG << "DBUS response error";
```
Comment (gtmills) [inline redfish-core/lib/environment_metrics.hpp:168]: Best if this was Error and included the ec

File: redfish-core/lib/environment_metrics.hpp [hunk 4]
```diff
@@ -152,6 +152,80 @@ inline void handleEnvironmentMetricsHead(
...
+            }
+
+            sdbusplus::asio::getProperty<double>(
+                *crow::connections::systemBus, object.begin()->first,
+                totalPowerPath, "xyz.openbmc_project.Sensor.Value", "Value",
+                [asyncResp, chassisId](const boost::system::error_code& ec2,
+                                       double value) {
+                if (ec2)
+                {
+                    BMCWEB_LOG_DEBUG << "Can't get Power Watts!";
```
Comment (gtmills) [inline redfish-core/lib/environment_metrics.hpp:211]: Best if this was Error and included the ec

File: redfish-core/lib/environment_metrics.hpp [hunk 5]
```diff
@@ -152,6 +152,80 @@ inline void handleEnvironmentMetricsHead(
...
+inline void getPowerWatts(const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+                          const std::string& chassisId)
+{
+    constexpr const char* interface = "xyz.openbmc_project.Sensor.Value";
+    const std::string sensorPath = "/xyz/openbmc_project/sensors";
+    constexpr std::array<std::string_view, 1> sensorInterfaces = {interface};
+    dbus::utility::getSubTreePaths(
+        sensorPath, 0, sensorInterfaces,
+        [asyncResp,
+         chassisId](const boost::system::error_code& ec,
```
Comment (gtmills) [inline redfish-core/lib/environment_metrics.hpp:164]: `#include <boost/system/error_code.hpp>`

**Review by gtmills — APPROVED** [review]
Defaulting to Chris here. Would like to see my minor comments addressed but not required

File: redfish-core/lib/environment_metrics.hpp (hunk 5 above)
Comment (lxwinspur) [reply redfish-core/lib/environment_metrics.hpp:166]: Done

**Comment by gtmills** [comment]
CI fail here: 
```
[242/243] Compiling C++ object bmcweb.p/src_webserver_main.cpp.o
FAILED: bmcweb.p/src_webserver_main.cpp.o 
c++ -Ibmcweb.p -I. -I.. -I../include -I../redfish-core/include -I../redfish-core/lib -I../http -Iconfig -I../config -flto=auto -fdiagnostics-color=never -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -Werror -std=c++20 -fno-rtti -O0 -g -fno-lto -DBMCWEB_ENABLE_BASIC_AUTHENTICATION -DBMCWEB_ENABLE_COOKIE_AUTHENTICATION -DBMCWEB_ENABLE_EVENT_SUBSCRIPTION_WEBSOCKET -DBMCWEB_ENABLE_HOST_SERIAL_WEBSOCKET -DBMCWEB_ENABLE_IBM_MANAGEMENT_CONSOLE -DBMCWEB_ENABLE_SSL -DBMCWEB_ENABLE_KVM -DBMCWEB_ENABLE_MUTUAL_TLS_AUTHENTICATION -DBMCWEB_ALLOW_DEPRECATED_POWER_THERMAL -DBMCWEB_ENABLE_REDFISH_BMC_JOURNAL -DBMCWEB_ENABLE_REDFISH_HOST_LOGGER -DBMCWEB_ENABLE_REDFISH_OEM_MANAGER_FAN_DATA -DBMCWEB_ENABLE_REDFISH_UPDATESERVICE_OLD_POST_URL -DBMCWEB_ENABLE_REDFISH -DBMCWEB_ENABLE_SESSION_AUTHENTICATION -DBMCWEB_ENABLE_STATIC_HOSTING -DBMCWEB_ENABLE_VM_WEBSOCKET -DBMCWEB_ENABLE_XTOKEN_AUTHENTICATION -Wcast-align -Wconversion -Wformat=2 -Wold-style-cast -Woverloaded-virtual -Wsign-conversion -Wunused -Wno-attributes -Wduplicated-cond -Wduplicated-branches -Wlogical-op -Wnull-dereference -Wunused-parameter -Wdouble-promotion -Wshadow -Wno-psabi -DBMCWEB_ENABLE_LOGGING -DBMCWEB_ENABLE_DEBUG -DBOOST_ALLOW_DEPRECATED_HEADERS -DBOOST_ASIO_NO_DEPRECATED -DBOOST_ASIO_SEPARATE_COMPILATION -DBOOST_BEAST_SEPARATE_COMPILATION -DBOOST_EXCEPTION_DISABLE -DBOOST_URL_NO_SOURCE_LOCATION -DJSON_NOEXCEPTION -DOPENSSL_NO_FILENAMES -DBOOST_ASIO_DISABLE_THREADS -DBOOST_ALL_NO_LIB -DBOOST_SYSTEM_NO_DEPRECATED -DBOOST_ERROR_CODE_HEADER_ONLY -DBOOST_COROUTINES_NO_DEPRECATION_WARNING -MD -MQ bmcweb.p/src_webserver_main.cpp.o -MF bmcweb.p/src_webserver_main.cpp.o.d -o bmcweb.p/src_webserver_main.cpp.o -c ../src/webserver_main.cpp
In file included from ../redfish-core/include/redfish.hpp:25,
                 from ../src/webserver_main.cpp:21:
../redfish-core/lib/environment_metrics.hpp: In lambda function:
../redfish-core/lib/environment_metrics.hpp:201:64: error: ‘ec’ is not captured
  201 |                 BMCWEB_LOG_ERROR << "DBUS response error: " << ec.message();
      |                                                                ^~
../redfish-core/lib/environment_metrics.hpp:197:28: note: the lambda has no capture-default
  197 |              totalPowerPath](const boost::system::error_code& ec1,
      |                            ^
../redfish-core/lib/environment_metrics.hpp:166:54: note: ‘const boost::system::error_code& ec’ declared here
  166 |          chassisId](const boost::system::error_code& ec,
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
../redfish-core/lib/environment_metrics.hpp: In lambda function:
../redfish-core/lib/environment_metrics.hpp:214:41: error: ‘ec’ is not captured
  214 |                                      << ec.message();
      |                                         ^~
../redfish-core/lib/environment_metrics.hpp:209:38: note: the lambda has no capture-default
  209 |                 [asyncResp, chassisId](const boost::system::error_code& ec2,
      |                                      ^
../redfish-core/lib/environment_metrics.hpp:166:54: note: ‘const boost::system::error_code& ec’ declared here
  166 |          chassisId](const boost::system::error_code& ec,
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
```

**Comment by lxwinspur** [comment]
Thanks @gtmills 
Updated

**Review by cjcain — APPROVED** [review]

**Pull request merged**